frame/election-provider-multi-phase/src/unsigned.rs

```diff
@@ -257,38 +260,36 @@ impl<T: Config> Pallet<T> {
 	/// If you want a checked solution and submit it at the same time, use
 	/// [`Pallet::mine_check_save_submit`].
 	pub fn mine_and_check(
-		iters: usize,
-	) -> Result<(RawSolution<SolutionOf<T>>, SolutionOrSnapshotSize), MinerError> {
-		let (raw_solution, witness) = Self::mine_solution(iters)?;
+	) -> Result<(RawSolution<SolutionOf<T>>, SolutionOrSnapshotSize), MinerError<T>> {
+		let (raw_solution, witness) = Self::mine_solution::<T::Solver>()?;
 		Self::basic_checks(&raw_solution, "mined")?;
 		Ok((raw_solution, witness))
 	}
 
 	/// Mine a new npos solution.
-	pub fn mine_solution(
-		iters: usize,
-	) -> Result<(RawSolution<SolutionOf<T>>, SolutionOrSnapshotSize), MinerError> {
+	/// In order to improve ergonomics, the Npos Solver type, `S`, must have the same
+	/// AccountId and Error type as the [`crate::Config::Solver`].
+	pub fn mine_solution<S>(
+	) -> Result<(RawSolution<SolutionOf<T>>, SolutionOrSnapshotSize), MinerError<T>>
+	where
+		S: NposSolver<AccountId = T::AccountId, Error = SolverErrorOf<T>>,
+	{
 		let RoundSnapshot { voters, targets } =
 			Self::snapshot().ok_or(MinerError::SnapshotUnAvailable)?;
 		let desired_targets = Self::desired_targets().ok_or(MinerError::SnapshotUnAvailable)?;
 
-		seq_phragmen::<_, SolutionAccuracyOf<T>>(
-			desired_targets as usize,
-			targets,
-			voters,
-			Some((iters, 0)),
-		)
-		.map_err(Into::into)
-		.and_then(Self::prepare_election_result)
+		S::solve(desired_targets as usize, targets, voters)
+			.map_err(|e| MinerError::Solver::<T>(e))
```

Comment:
```suggestion
			.map_err(|e| MinerError::Solver::<T>(e))
```

I think the compiler forced us to do it this way, maybe it is now time to revisit why `From` doesn't work?

Reply:
I get the below error; I am guessing the issue is that `<<T as Config>::Solver as NposSolver>::Error` (aka `SolverErrorOf<T>`) could technically be `MinerError` .. but I am not sure if that really explains the compiler error

```rust
  Compiling pallet-election-provider-multi-phase v4.0.0-dev (/home/emostov/remote-builds/substrate-review/frame/election-provider-multi-phase)
error[E0119]: conflicting implementations of trait `std::convert::From<unsigned::MinerError<_>>` for type `unsigned::MinerError<_>`
   --> frame/election-provider-multi-phase/src/unsigned.rs:105:1
    |
105 | impl<T: Config> From<SolverErrorOf<T>> for MinerError<T> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: conflicting implementation in crate `core`:
            - impl<T> From<T> for T;

error: aborting due to previous error

For more information about this error, try `rustc --explain E0119`.
error: could not compile `pallet-election-provider-multi-phase`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error[E0119]: conflicting implementations of trait `std::convert::From<unsigned::MinerError<_>>` for type `unsigned::MinerError<_>`
   --> frame/election-provider-multi-phase/src/unsigned.rs:105:1
    |
105 | impl<T: Config> From<SolverErrorOf<T>> for MinerError<T> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: conflicting implementation in crate `core`:
            - impl<T> From<T> for T;

error: aborting due to previous error

For more information about this error, try `rustc --explain E0119`.
```

Reply:
agg bugger me, yeah it does make sense. We explicitely constrain somewhere in the code that `T: Solver<Error = MinerError>`, so basically `SolverErrorOf<T> == MinerError<T>`.